partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/ExcludeUtils.kt

```diff
@@ -0,0 +1,159 @@
+                            }
+                        }
+                        is Identifier.Qualified -> {
+                            if (id.root.isEquivalentTo(binding.name)) {
+                                binding.type.checkExclude(steps)
+                            } else {
+                                false
+                            }
+                        }
+                    }
+                }
+                is Rex.Op.Var.Resolved -> false
+            }
+        }
+        // If nothing is excluded by `excludePath`, add a warning
+        if (!excludesSomething) {
+            onProblem(
+                Problem(
+                    sourceLocation = UNKNOWN_PROBLEM_LOCATION,
+                    details = PlanningProblemDetails.InvalidExcludePath(
+                        excludePath.toProblemString()
+                    )
+                )
+            )
+        }
+    }
+}
+
+/**
+ * Checks whether [steps] will exclude a value from [this] [StaticType].
+ */
+internal fun StaticType.checkExclude(steps: List<Rel.Op.Exclude.Step>): Boolean {
+    return when (this) {
+        is StructType -> this.checkExclude(steps)
+        is CollectionType -> this.checkExclude(steps)
+        is AnyOfType -> this.types.any { it.checkExclude(steps) }
+        else -> false
+    }
+}
+
+/**
+ * Checks whether [steps] will exclude a value from [this] [StructType].
+ */
+internal fun StructType.checkExclude(steps: List<Rel.Op.Exclude.Step>): Boolean {
+    // Ignore open structs
+    if (!this.contentClosed) {
+        return true
+    }
+    val step = steps.first()
+    return fields.any { field ->
+        if (steps.size == 1) {
+            when (step) {
+                is Rel.Op.Exclude.Step.StructField -> {
+                    step.symbol.isEquivalentTo(field.key)
+                }
+                is Rel.Op.Exclude.Step.StructWildcard -> true
+                else -> false
+            }
+        } else {
+            when (step) {
+                is Rel.Op.Exclude.Step.StructField -> {
+                    step.symbol.isEquivalentTo(field.key) && field.value.checkExclude(steps.drop(1))
+                }
+                is Rel.Op.Exclude.Step.StructWildcard -> true
+                else -> false
+            }
+        }
+    }
+}
+
+/**
+ * Checks whether [steps] will exclude a value from [this] [CollectionType].
+ */
+internal fun CollectionType.checkExclude(steps: List<Rel.Op.Exclude.Step>): Boolean {
+    val step = steps.first()
+    return if (steps.size == 1) {
+        when (step) {
+            is Rel.Op.Exclude.Step.CollIndex, is Rel.Op.Exclude.Step.CollWildcard -> true
+            else -> false
+        }
+    } else {
+        when (step) {
+            is Rel.Op.Exclude.Step.CollIndex, is Rel.Op.Exclude.Step.CollWildcard -> {
+                val e = this.elementType
+                e.checkExclude(steps.drop(1))
+            }
+            else -> false
+        }
+    }
+}
```

Comment:
You could actually simplify these to something like

```kotlin
if (steps.isEmpty()) {
    return false
}
when (steps.first()) {
            is Rel.Op.Exclude.Step.CollIndex, is Rel.Op.Exclude.Step.CollWildcard -> {
                val e = this.elementType
                e.checkExclude(steps.drop(1))
            }
            else -> false
        }
}
```

I don't believe you need the if size thing because steps.drop(1) will return an empty list and then your recursive base case hits.

Reply:
Thanks for the suggestion. I simplified the `checkExclude` for structs and collection types based on your suggestion in the latest commit.